### ISSUE TITLE
[Docs/Feature] Extend the deployment lifecycle with app termination and nginx proxy steps

### ISSUE THREAD

**Issue by lambrospetrou** [aws/elastic-beanstalk-roadmap#222]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or "me too" comments. They generate extra noise for issue followers and do not help prioritize the request.

**Tell us about your request**
What do you want us to build?

I would like to see some updates in the [Extending Linux platforms](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/platforms-linux-extend.html) documentation. It nicely describes the [instance deployment lifecycle](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/platforms-linux-extend.html#platforms-linux-extend.workflow), especially when it comes to platform/configuration hooks.

However, there are some things missing for which I could not find any documentation anywhere (and Google does not help either):
1. How does the termination of the running application happens. E.g. I have 2 processes in my `Procfile`. When do they get shutdown in the instance lifecycle (e.g. after or before the `predeploy` step?)?
2. It's not clear from the diagram what happens with the `nginx` proxy. Are you doing `nginx reload` or stop/start? I am investigating all the steps that essentially cause downtime during deployments and one key aspect of it is the duration when the `nginx` proxy is down not accepting requests.
3. Is there a way to hook into the main `systemd` flow coordinating the deployment, or is it only possible to add my own `systemctl` commands inside `.platform/hooks`?
4. Is there a way to disable `nginx` for the Go platform, similar to the Docker and [.NET core platforms](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/command-options-specific.html#command-options-dotnet-core-linux) (i.e. `aws:elasticbeanstalk:environment:proxy`? Not having `nginx` would make it possible to use [`systemd` socket activation to essentially enable **zero downtime** during deployments](https://github.com/lambrospetrou/aws-playground/tree/master/elastic-beanstalk-zero-downtime-systemd/golang). Is there a way to achieve this with the current nginx setup?

**Is this request specific to an Elastic Beanstalk [platform](https://docs.aws.amazon.com/elasticbeanstalk/latest/platforms/platforms-supported.html)?**
If so, which one(s)?

No, all AL2 Linux platforms. Except the option to disable `nginx` for Go platform.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
What outcome are you trying to achieve, ultimately, and why is it hard/impossible to do right now? What is the impact of not having this problem solved? The more details you can provide, the better we'll be able to understand and solve the problem.

1. I want to have a more detailed view on what happens during a deployment (code, or configuration) with the built-in components, especially the proxy, and the app termination (when it happens, and how, e.g. signals?).
2. I want to understand what can be done (in an easy or complicated way) to achieve **zero downtime single-instance** deployments.

**Are you currently working around this issue?**
How are you currently solving this problem?

- Without knowing specifics about how the `nginx` proxy is managed it's hard to workaround. I am trying to reverse engineer it through logs but official documentation would be much better :)

**Additional context**
Anything else we should know?

**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

**Comment by zzzhu0192** [comment]
Hi,

We will look into updating our documentation.

As for your questions, here are the answers. 

* > How does the termination of the running application happens. E.g. I have 2 processes in my Procfile. When do they get shutdown in the instance lifecycle (e.g. after or before the predeploy step?)?

Processes are shut down after predeploy step.

* > It's not clear from the diagram what happens with the nginx proxy. Are you doing nginx reload or stop/start? I am investigating all the steps that essentially cause downtime during deployments and one key aspect of it is the duration when the nginx proxy is down not accepting requests.

We are doing nginx stop/start instead reload.

* > Is there a way to hook into the main systemd flow coordinating the deployment, or is it only possible to add my own systemctl commands inside .platform/hooks?

Can you elaborate on what kinds of customization you need ? And by main systemd flow, are you referring to `web.service` ? We have other customers asking for more systemd customization as well. e.g. allowing more customization than `ExecStart` command. It could be helpful for us to understand different use cases. 

* > Is there a way to disable nginx for the Go platform, similar to the Docker and [.NET core platforms](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/command-options-specific.html#command-options-dotnet-core-linux) (i.e. aws:elasticbeanstalk:environment:proxy? Not having nginx would make it possible to use [systemd socket activation to essentially enable zero downtime during deployments](https://github.com/lambrospetrou/aws-playground/tree/master/elastic-beanstalk-zero-downtime-systemd/golang). Is there a way to achieve this with the current nginx setup?

Our Go platform does not support none proxy by default. I cannot think of an easy way to bypass nginx related instruction with current setup. We will consider add none proxy support for Go platform. 

Thanks

**Comment by lambrospetrou** [comment]
@zzzhu0192 Thanks for looking into this. Updates in the documentation to include termination and nginx details will be fantastic.
I pretty much reversed engineered it now by logging in every step, sshing into the instance and tracking the `eb-engine.log` mostly but official specs will be much better.

>Can you elaborate on what kinds of customization you need ? And by main systemd flow, are you referring to web.service ? We have other customers asking for more systemd customization as well. e.g. allowing more customization than ExecStart command. It could be helpful for us to understand different use cases.

For example, if I want to add my own systemd services/sockets/whatever scripts at the moment it's not very clear what step is the right hook to use. If I want to add something in systemd and start it before the main app (Procfile `web` process) but from inside the final directory (`/var/current/app`) it's not possible at the moment. It is either from predeploy hook in the staging directory, or in the postdeploy hook after the app started.

In general, the platform hooks are easy but the only way to be certain we are inside the final directory is in the `postdeploy` which might be too late for certain things.

>Our Go platform does not support none proxy by default. I cannot think of an easy way to bypass nginx related instruction with current setup. We will consider add none proxy support for Go platform.

I did some hacking on this and it's possible but very cumbersome. I can either completely disable nginx by providing an empty configuration or simply change its port to something other than `80`, which will allow my web service to use that port directly.

Also, the systemd limitations mentioned above, make it very ugly to use systemd socket activation right now because there is no hook to run from inside the final directory before the app starts. I managed to do it (will write a blog post soon) but it needs some more conveniences.

TL;DR: What I want to do essentially is use the Elastic Beanstalk Single Instance environment which is a managed platform (OS updates, patches, etc) but use `systemd` to manage my application through socket activation instead of the normal start/stop flow which causes downtime.

Thanks
